### ISSUE TITLE
Problem with Testing

### ISSUE THREAD

**Issue by Subash33** [mravanelli/pytorch-kaldi#113]
Hello,

First of all thank you guys for such an amazing toolkit. I have been trying to use it for my project. Just to see it works, I ran the experiment for one epoch and tried with different number of test and dev chunks, 1 or 4. My issue is with testing phase. It throws out the out the list index of range error. I was able to run the TIMIT database experiment earlier and now using TIMIT  MLP basic and get the results. The path is correct, the required kaldi files are in place, 
![pytorch_error](https://user-images.githubusercontent.com/44344741/59805306-e7afa400-92be-11e9-8f92-ae74e7c06c8e.png)
I don't understand how exactly can I fix this for my dataset. I am attaching the conf file and the error output.

Your help will be highly appreciated.
-Subash

"Config file"

[cfg_proto]
cfg_proto = proto/global.proto
cfg_proto_chunk = proto/global_chunk.proto

[exp]
cmd = 
run_nn_script = run_nn
out_folder = /home/subash/EMA_MAE/pytorch-exps/Acoustic/MLP
seed = 1234
use_cuda = False
multi_gpu = False
save_gpumem = True
n_epochs_tr = 1

[dataset1]
data_name = EMAMAE_tr
fea = fea_name=featdir
	fea_lst=/home/subash/kaldi/egs/Acoustic/data/train/feats.scp
	fea_opts=apply-cmvn --utt2spk=ark:/home/subash/kaldi/egs/Acoustic/data/train/utt2spk  ark:/home/subash/kaldi/egs/Acoustic/data/train/data/cmvn_train.ark ark:- ark:- | 
	cw_left=5
	cw_right=5
	

lab = lab_name=lab_cd
	lab_folder=/home/subash/kaldi/egs/Acoustic/exp/tri3_ali
	lab_opts=ali-to-pdf
	lab_count_file=auto
	lab_data_folder=/home/subash/kaldi/egs/Acoustic/data/train/
	lab_graph=/home/subash/kaldi/egs/Acoustic/exp/tri3/graph
	

n_chunks = 5

[dataset2]
data_name = EMAMAE_dev
fea = fea_name=featdir
	fea_lst=/home/subash/kaldi/egs/Acoustic/data/dev/feats.scp
	fea_opts=apply-cmvn --utt2spk=ark:/home/subash/kaldi/egs/Acoustic/data/dev/utt2spk  ark:/home/subash/kaldi/egs/Acoustic/data/dev/data/cmvn_dev.ark ark:- ark:- | 
	cw_left=5
	cw_right=5
	

lab = lab_name=lab_cd
	lab_folder=/home/subash/kaldi/egs/Acoustic/exp/tri3_ali_dev
	lab_opts=ali-to-pdf
	lab_count_file=auto
	lab_data_folder=/home/subash/kaldi/egs/Acoustic/data/dev/
	lab_graph=/home/subash/kaldi/egs/Acoustic/exp/tri3/graph
	

n_chunks = 4

[dataset3]
data_name = EMAMAE_test

fea = fea_name=featdir
	
	fea_lst=/home/subash/kaldi/egs/Acoustic/data/test/feats.scp
	fea_opts=apply-cmvn --utt2spk=ark:/home/subash/kaldi/egs/Acoustic/data/test/utt2spk  ark:/home/subash/kaldi/egs/Acoustic/data/test/data/cmvn_test.ark ark:- ark:- | 
	cw_left=5
	cw_right=5
	

lab = lab_name=lab_cd
	lab_folder=/home/subash/kaldi/egs/Acoustic/exp/tri3_ali_test
	lab_opts=ali-to-pdf
	lab_count_file=auto
	lab_data_folder=/home/subash/kaldi/egs/Acoustic/data/test/
	lab_graph=/home/subash/kaldi/egs/Acoustic/exp/tri3/graph
	
	

n_chunks = 4

[data_use]
train_with = EMAMAE_tr
valid_with = EMAMAE_dev
forward_with = EMAMAE_test

[batches]
batch_size_train = 128
max_seq_length_train = 1000
increase_seq_length_train = False
start_seq_len_train = 100
multply_factor_seq_len_train = 2
batch_size_valid = 128
max_seq_length_valid = 1000

[architecture1]
arch_name = MLP_layers1
arch_proto = proto/MLP.proto
arch_library = neural_networks
arch_class = MLP
arch_pretrain_file = none
arch_freeze = False
arch_seq_model = False
dnn_lay = 1024,1024,1024,1024,N_out_lab_cd
dnn_drop = 0.15,0.15,0.15,0.15,0.0
dnn_use_laynorm_inp = False
dnn_use_batchnorm_inp = False
dnn_use_batchnorm = True,True,True,True,False
dnn_use_laynorm = False,False,False,False,False
dnn_act = relu,relu,relu,relu,softmax
arch_lr = 0.08
arch_halving_factor = 0.5
arch_improvement_threshold = 0.001
arch_opt = sgd
opt_momentum = 0.0
opt_weight_decay = 0.0
opt_dampening = 0.0
opt_nesterov = False

[model]
model_proto = proto/model.proto
model = out_dnn1=compute(MLP_layers1,featdir)
	loss_final=cost_nll(out_dnn1,lab_cd)
	err_final=cost_err(out_dnn1,lab_cd)

[forward]
forward_out = out_dnn1
normalize_posteriors = True
normalize_with_counts_from = lab_cd
save_out_file = False
require_decoding = True

[decoding]
decoding_script_folder = kaldi_decoding_scripts/
decoding_script = decode_dnn.sh
decoding_proto = proto/decoding.proto
min_active = 200
max_active = 7000
max_mem = 50000000
beam = 13.0
latbeam = 8.0
acwt = 0.2
max_arcs = -1
skip_scoring = false
scoring_script = local/score.sh
scoring_opts = "--min-lmwt 1 --max-lmwt 10"
norm_vars = False

**Comment by mravanelli** [comment]
Hi,
thank you very much for your interest in our toolkit. That error happens
when the system is not to able to correctly read the features or the labels
(in the new version that we are devoloping these days this error will be
written in a much more clear way). You can take a look into some previous
issue such as https://github.com/mravanelli/pytorch-kaldi/issues/105
Typically,  it could be important to do some prints within the load_dataset
funcction in data_io.py. This is the function that reads the features and
the labels and one might first of all take a look into the fea and lab
dictionaries and see if they are empty or not.

Best,

Mirco

**Comment by weienzhang** [comment]
> Hello,
> 
> First of all thank you guys for such an amazing toolkit. I have been trying to use it for my project. Just to see it works, I ran the experiment for one epoch and tried with different number of test and dev chunks, 1 or 4. My issue is with testing phase. It throws out the out the list index of range error. I was able to run the TIMIT database experiment earlier and now using TIMIT MLP basic and get the results. The path is correct, the required kaldi files are in place,
> ![pytorch_error](https://user-images.githubusercontent.com/44344741/59805306-e7afa400-92be-11e9-8f92-ae74e7c06c8e.png)
> I don't understand how exactly can I fix this for my dataset. I am attaching the conf file and the error output.
> 
> Your help will be highly appreciated.
> -Subash
> 
> "Config file"
> 
> [cfg_proto]
> cfg_proto = proto/global.proto
> cfg_proto_chunk = proto/global_chunk.proto
> 
> [exp]
> cmd =
> run_nn_script = run_nn
> out_folder = /home/subash/EMA_MAE/pytorch-exps/Acoustic/MLP
> seed = 1234
> use_cuda = False
> multi_gpu = False
> save_gpumem = True
> n_epochs_tr = 1
> 
> [dataset1]
> data_name = EMAMAE_tr
> fea = fea_name=featdir
> fea_lst=/home/subash/kaldi/egs/Acoustic/data/train/feats.scp
> fea_opts=apply-cmvn --utt2spk=ark:/home/subash/kaldi/egs/Acoustic/data/train/utt2spk ark:/home/subash/kaldi/egs/Acoustic/data/train/data/cmvn_train.ark ark:- ark:- |
> cw_left=5
> cw_right=5
> 
> lab = lab_name=lab_cd
> lab_folder=/home/subash/kaldi/egs/Acoustic/exp/tri3_ali
> lab_opts=ali-to-pdf
> lab_count_file=auto
> lab_data_folder=/home/subash/kaldi/egs/Acoustic/data/train/
> lab_graph=/home/subash/kaldi/egs/Acoustic/exp/tri3/graph
> 
> n_chunks = 5
> 
> [dataset2]
> data_name = EMAMAE_dev
> fea = fea_name=featdir
> fea_lst=/home/subash/kaldi/egs/Acoustic/data/dev/feats.scp
> fea_opts=apply-cmvn --utt2spk=ark:/home/subash/kaldi/egs/Acoustic/data/dev/utt2spk ark:/home/subash/kaldi/egs/Acoustic/data/dev/data/cmvn_dev.ark ark:- ark:- |
> cw_left=5
> cw_right=5
> 
> lab = lab_name=lab_cd
> lab_folder=/home/subash/kaldi/egs/Acoustic/exp/tri3_ali_dev
> lab_opts=ali-to-pdf
> lab_count_file=auto
> lab_data_folder=/home/subash/kaldi/egs/Acoustic/data/dev/
> lab_graph=/home/subash/kaldi/egs/Acoustic/exp/tri3/graph
> 
> n_chunks = 4
> 
> [dataset3]
> data_name = EMAMAE_test
> 
> fea = fea_name=featdir
> 
> ```
> fea_lst=/home/subash/kaldi/egs/Acoustic/data/test/feats.scp
> fea_opts=apply-cmvn --utt2spk=ark:/home/subash/kaldi/egs/Acoustic/data/test/utt2spk  ark:/home/subash/kaldi/egs/Acoustic/data/test/data/cmvn_test.ark ark:- ark:- | 
> cw_left=5
> cw_right=5
> ```
> 
> lab = lab_name=lab_cd
> lab_folder=/home/subash/kaldi/egs/Acoustic/exp/tri3_ali_test
> lab_opts=ali-to-pdf
> lab_count_file=auto
> lab_data_folder=/home/subash/kaldi/egs/Acoustic/data/test/
> lab_graph=/home/subash/kaldi/egs/Acoustic/exp/tri3/graph
> 
> n_chunks = 4
> 
> [data_use]
> train_with = EMAMAE_tr
> valid_with = EMAMAE_dev
> forward_with = EMAMAE_test
> 
> [batches]
> batch_size_train = 128
> max_seq_length_train = 1000
> increase_seq_length_train = False
> start_seq_len_train = 100
> multply_factor_seq_len_train = 2
> batch_size_valid = 128
> max_seq_length_valid = 1000
> 
> [architecture1]
> arch_name = MLP_layers1
> arch_proto = proto/MLP.proto
> arch_library = neural_networks
> arch_class = MLP
> arch_pretrain_file = none
> arch_freeze = False
> arch_seq_model = False
> dnn_lay = 1024,1024,1024,1024,N_out_lab_cd
> dnn_drop = 0.15,0.15,0.15,0.15,0.0
> dnn_use_laynorm_inp = False
> dnn_use_batchnorm_inp = False
> dnn_use_batchnorm = True,True,True,True,False
> dnn_use_laynorm = False,False,False,False,False
> dnn_act = relu,relu,relu,relu,softmax
> arch_lr = 0.08
> arch_halving_factor = 0.5
> arch_improvement_threshold = 0.001
> arch_opt = sgd
> opt_momentum = 0.0
> opt_weight_decay = 0.0
> opt_dampening = 0.0
> opt_nesterov = False
> 
> [model]
> model_proto = proto/model.proto
> model = out_dnn1=compute(MLP_layers1,featdir)
> loss_final=cost_nll(out_dnn1,lab_cd)
> err_final=cost_err(out_dnn1,lab_cd)
> 
> [forward]
> forward_out = out_dnn1
> normalize_posteriors = True
> normalize_with_counts_from = lab_cd
> save_out_file = False
> require_decoding = True
> 
> [decoding]
> decoding_script_folder = kaldi_decoding_scripts/
> decoding_script = decode_dnn.sh
> decoding_proto = proto/decoding.proto
> min_active = 200
> max_active = 7000
> max_mem = 50000000
> beam = 13.0
> latbeam = 8.0
> acwt = 0.2
> max_arcs = -1
> skip_scoring = false
> scoring_script = local/score.sh
> scoring_opts = "--min-lmwt 1 --max-lmwt 10"
> norm_vars = False

hello ! May i know that was your problem solved ? And i just has the similar question when i ran TIMIT_liGRU_fmllr and it went wrong in the first chunk of train data...
